src/Features/Core/Portable/EmbeddedLanguages/AbstractLanguageDetector.cs

```diff
@@ -0,0 +1,173 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.VirtualChars;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Features.EmbeddedLanguages
+{
+    internal abstract class AbstractLanguageDetector<TOptions, TTree>
+        where TOptions : struct, Enum
+        where TTree : class
+    {
+        protected readonly EmbeddedLanguageInfo Info;
+        private readonly LanguageCommentDetector<TOptions> _commentDetector;
+
+        protected AbstractLanguageDetector(
+            EmbeddedLanguageInfo info,
+            LanguageCommentDetector<TOptions> commentDetector)
+        {
+            Info = info;
+            _commentDetector = commentDetector;
+        }
+
+        protected abstract bool IsEmbeddedLanguageString(SyntaxToken token, SyntaxNode argumentNode, SemanticModel semanticModel, CancellationToken cancellationToken, out TOptions options);
+        protected abstract TTree? TryParse(VirtualCharSequence chars, TOptions options);
+        protected abstract bool TryGetOptions(SemanticModel semanticModel, SyntaxNode expr, TypeInfo exprType, CancellationToken cancellationToken, out TOptions options);
+
+        public bool IsPossiblyPatternToken(SyntaxToken token, ISyntaxFacts syntaxFacts)
+        {
+            if (!syntaxFacts.IsStringLiteral(token))
+                return false;
+
+            if (syntaxFacts.IsLiteralExpression(token.Parent) && syntaxFacts.IsArgument(token.Parent.Parent))
+                return true;
+
+            return HasLanguageComment(token, syntaxFacts, out _);
+        }
+
+        private bool HasLanguageComment(
+            SyntaxToken token, ISyntaxFacts syntaxFacts, out TOptions options)
+        {
+            if (HasLanguageComment(token.GetPreviousToken().TrailingTrivia, syntaxFacts, out options))
+                return true;
+
+            for (var node = token.Parent; node != null; node = node.Parent)
+            {
+                if (HasLanguageComment(node.GetLeadingTrivia(), syntaxFacts, out options))
+                    return true;
+
+                // Stop walking up once we hit a statement.  We don't need/want statements higher up the parent chain to
+                // have any impact on this token.
+                if (syntaxFacts.IsStatement(node))
+                    break;
```

Comment:
(I recognized this may have moved, asking anyways):

Does this need some check for lambdas, or OK with something outside a lambda impacting inside of it? I think there's a reasonable argument for it, this just my general caution any time I see check similar to this.

Reply:
yeah.  i think this is acceptable for now.  We could always grow this later if users find this undesirable.